File: mobiletrm-backend/src/main/java/world/inetumrealdolmen/dto/CustomerDto.java
package world.inetumrealdolmen.dto;

import world.inetumrealdolmen.domain.Customer;

/**
 * A collection of Data Transfer Objects (DTO) for {@link Customer} entities to be sent to clients.
 * Use the static subclasses' constructors to create DTO's.
 */
public class CustomerDto {

  /**
   * Represents an index view of a customer, containing information for a project's details.
   */
  public static class Index {

    /**
     * The ID of the customer.
     */
    public long id;

    /**
     * The name of the customer.
     */
    public String name;

    /**
     * The name of the company the customer is associated with.
     */
    public String companyName;

    /**
     * Default constructor for Jackson serializing.
     * Protected to prevent instantiation without parameters.
     */
    protected Index() {
    }

    /**
     * Constructor to convert a Customer object into a DTO.
     *
     * @param customer The Customer object to convert.
     */
    public Index(Customer customer) {
      this.id = customer.id;
      this.name = customer.name;
      this.companyName = customer.companyName;
    }
  }
}

File: mobiletrm-android/app/src/main/java/world/inetumrealdolmen/mobiletrm/data/dto/CustomerDTO.kt
package world.inetumrealdolmen.mobiletrm.data.dto

import kotlinx.serialization.Serializable
import world.inetumrealdolmen.mobiletrm.data.model.Customer

/**
 * Index information about the customer from the api.
 *
 * @param id The database ID of the customer.
 * @param name The full name of the customer.
 * @param companyName The name of the company that employs the customer.
 */
@Serializable
data class CustomerDTO(
    val id: Long,
    val name: String,
    val companyName: String,
)

/**
 * Maps a Customer received from the API to a model object.
 *
 * @return The model object of the Customer.
 */
fun CustomerDTO.asDomainObject() =
    Customer(
        id = id,
        name = name,
        companyName = companyName,
    )

/**
 * Maps a Customer modele to a DTO.
 *
 * @return The DTO of the Customer.
 */
fun Customer.asDto() =
    CustomerDTO(
        id = id,
        name = name,
        companyName = companyName,
    )
